x/fswap/keeper/keeper.go

```diff
@@ -82,6 +105,63 @@ func (k Keeper) Swap(ctx sdk.Context, addr sdk.AccAddress, fromCoinAmount sdk.Co
 	return nil
 }
 
+func (k Keeper) SetSwap(ctx sdk.Context, swap types.Swap, toDenomMetadata bank.Metadata) error {
+	isNewSwap := true
+	if _, err := k.getSwap(ctx, swap.FromDenom, swap.ToDenom); err == nil {
+		isNewSwap = false
+	}
+
+	if !isNewSwap && !k.config.UpdateAllowed {
+		return sdkerrors.ErrInvalidRequest.Wrap("update existing swap not allowed")
+	}
+
+	if isNewSwap {
+		if err := k.increaseSwapCount(ctx); err != nil {
+			return err
+		}
+	}
+
+	stats, err := k.getSwapStats(ctx)
+	if err != nil {
+		return err
+	}
+
+	if int(stats.SwapCount) > k.config.MaxSwaps && !k.isUnlimited() {
+		return types.ErrCanNotHaveMoreSwap.Wrapf("cannot make more swaps, max swaps is %d", k.config.MaxSwaps)
+	}
+
```

Comment:
Is it necessary to add a logic validation to validate whether from-denom is exist on chain?

Reply:
`isNewSwap` is check flag for existing swap.
Current design allow updating only `Swap` up to `Config`.

What is purpose of logic validation? could you share example use-case you have?

Reply:
https://github.com/Finschia/finschia-sdk/commit/2ef1127d86f022a1a4e1920d54fc12b50a470b3e#diff-358b36b4874c98e18bbc5bbefbea63a95236611fd8c2f417e397bee0e7a855cdR40-R42

Reply:
> What is purpose of logic validation? could you share example use-case you have?

Because when I tried before, there was only `stake` on the chain, but it passed the setSwap requirement with a fromdenom of `cony`.

Reply:
It looks good to introduce `HasSupply(ctx, swap.FromDenom)`. But it's not related to this PR. We may apply it with another PR.

Reply:
I will prepare a pr to refactor ut tests and this point.